fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -823,6 +828,92 @@ public void buildIndex() {
+     * @param buildRangeProcessor build index for entries between given the primary key endpoints (first inclusive,
+     *     second exclusive) and mark the index as readable when all subtasks are done
+     */
+    @API(API.Status.EXPERIMENTAL)
+    @Nonnull
+    public void buildIndexInMultipleJobs(int minParallelism, int maxParallelism,
+                                         Consumer<Pair<Tuple, Tuple>> buildRangeProcessor) {
+        TupleRange tupleRange = runner.asyncToSync(FDBStoreTimer.Waits.WAIT_BUILD_ENDPOINTS, buildEndpoints());
+
+        if (tupleRange != null) {
+            List<Tuple> boundaries = getPrimaryKeyBoundaries(tupleRange);
+            if (boundaries.size() - 1 < minParallelism) {
+                asyncToSync(
+                        buildRange(Key.Evaluated.fromTuple(tupleRange.getLow()), Key.Evaluated.fromTuple(tupleRange.getHigh())));
+            } else {
+                List<Pair<Tuple, Tuple>> tasks = new ArrayList<>();
+                int stepSize = (boundaries.size() + maxParallelism - 1) / maxParallelism; // step size >= 1
+                int start = 0;
+                while (true) {
+                    int next = start + stepSize;
+                    if (next < boundaries.size() - 1) {
+                        tasks.add(Pair.of(boundaries.get(start), boundaries.get(next)));
+                    } else {
+                        tasks.add(Pair.of(boundaries.get(start), boundaries.get(boundaries.size() - 1)));
+                        break;
+                    }
+                }
+                // Shuffle the tasks to avoid hot spots.
+                Collections.shuffle(tasks);
+                LOGGER.info(KeyValueLogMessage.of("build index in parallel",
+                        "indexName", index.getName(),
+                        "ranges", tasks));
+                tasks.forEach(buildRangeProcessor);
+                // The index should not be marked as readable. But remember to do it when all subtasks are done.
+                return;
+            }
+        }
+        runner.run(context -> recordStoreBuilder.copyBuilder().setContext(context).open().markIndexReadable(index));
+    }
+
+    private List<Tuple> getPrimaryKeyBoundaries(TupleRange tupleRange) {
+        List<Tuple> boundaries = runner.run(context -> {
+            RecordCursor<Tuple> cursor = recordStoreBuilder.copyBuilder().setContext(context).open()
+                    .getPrimaryKeyBoundaries(tupleRange.getLow(), tupleRange.getHigh());
+            List<Tuple> keys = new ArrayList<>();
+            cursor.forEachRemaining(keys::add);
+            return keys;
+        });
+
+        // Add the two endpoints if they are not in the result
+        if (!boundaries.get(0).equals(tupleRange.getLow())) {
+            boundaries.add(0, tupleRange.getLow());
+        }
+        if (!boundaries.get(boundaries.size() - 1).equals(tupleRange.getHigh())) {
+            boundaries.add(tupleRange.getHigh());
+        }
```

Comment:
If the assumption is that the tuples returned are ordered, it would probably make more sense to check something like:

    if (tupleRange.getLow().compareTo(boundaries.get(0)) < 0) {
        boundaries.add(0, tupleRange.getLow());
    }
    if (tupleRange.getHigh().compareTo(boundaries.get(boundaries.size() - 1) > 0) {
         boundaries.add(0, tupleRange.getHigh());
    }